Deprecated std::iterator

As initially reported [here](https://github.com/RcppCore/RcppParallel/issues/191#issuecomment-1407462280):

```
* checking whether package 'detrendr' can be installed ... WARNING
Found the following significant warnings:
  D:/RCompile/CRANpkg/lib/4.3/RcppParallel/include/tbb/concurrent_hash_map.h:343:23: warning: 'template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator' is deprecated [-Wdeprecated-declarations]
  D:/RCompile/CRANpkg/lib/4.3/RcppParallel/include/tbb/internal/_concurrent_queue_impl.h:744:21: warning: 'template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator' is deprecated [-Wdeprecated-declarations]
  D:/RCompile/CRANpkg/lib/4.3/RcppParallel/include/tbb/internal/_concurrent_queue_impl.h:1003:21: warning: 'template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator' is deprecated [-Wdeprecated-declarations]
  D:/RCompile/CRANpkg/lib/4.3/RcppParallel/include/tbb/internal/_concurrent_unordered_impl.h:67:36: warning: 'template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator' is deprecated [-Wdeprecated-declarations]
  D:/RCompile/CRANpkg/lib/4.3/RcppParallel/include/RcppParallel/RMatrix.h:18:24: warning: 'template<class _Category, class _Tp, class _Distance, class _Pointer, class _Reference> struct std::iterator' is deprecated [-Wdeprecated-declarations]
```

`std::iterator` has been deprecated, so we should move away from it.

--- a/NEWS.md
+++ b/NEWS.md
@@ -1,6 +1,8 @@
 
 ## RcppParallel 5.1.7 (UNRELEASED)
 
+* Remove deprecated `std::iterator`. (#192; @Enchufa2)
+
 ## RcppParallel 5.1.6
 
 * Patch for TBB to allow compilation with gcc-13.

--- a/inst/include/RcppParallel/RMatrix.h
+++ b/inst/include/RcppParallel/RMatrix.h
@@ -14,16 +14,21 @@ public:
    public:   
       
       template <typename V>
-      class row_iterator 
-         : public std::iterator<std::random_access_iterator_tag, V, std::size_t> {
+      class row_iterator {
       
       public:
-         inline row_iterator(Row& row, std::size_t i)
+         using iterator_category = std::random_access_iterator_tag;
+         using value_type = V;
+         using difference_type = std::size_t;
+         using pointer = value_type*;
+         using reference = value_type&;
+
+         inline row_iterator(Row& row, difference_type i)
             : start_(row.start_), parentNrow_(row.parent_.nrow()), index_(i)
          {
          }
          
-         inline row_iterator(V* start, std::size_t parentNrow, std::size_t index)
+         inline row_iterator(pointer start, difference_type parentNrow, difference_type index)
             : start_(start), parentNrow_(parentNrow), index_(index)
          {      
          }
@@ -57,23 +62,23 @@ public:
             return tmp ;
          }
 
-         row_iterator operator+(std::size_t n) const { 
+         row_iterator operator+(difference_type n) const {
             return row_iterator(start_, parentNrow_ ,index_ + n ) ; 
          }
-         row_iterator operator-(std::size_t n) const { 
+         row_iterator operator-(difference_type n) const {
             return row_iterator(start_, parentNrow_, index_ - n ) ; 
          }
          
-         std::size_t operator+(const row_iterator& other) const {
+         difference_type operator+(const row_iterator& other) const {
             return index_ + other.index_;
          }
          
-         std::size_t operator-(const row_iterator& other) const { 
+         difference_type operator-(const row_iterator& other) const {
             return index_ - other.index_ ; 
          }
          
-         row_iterator& operator+=(std::size_t n) { index_ += n ; return *this; }
-         row_iterator& operator-=(std::size_t n) { index_ -= n ; return *this; }
+         row_iterator& operator+=(difference_type n) { index_ += n ; return *this; }
+         row_iterator& operator-=(difference_type n) { index_ -= n ; return *this; }
          
          bool operator==(const row_iterator& other) const { return index_ == other.index_; }
          bool operator!=(const row_iterator& other) const { return index_ != other.index_; }
@@ -83,16 +88,16 @@ public:
          bool operator>=(const row_iterator& other) const { return index_ >= other.index_; }
          
 
-         inline V& operator*() { return start_[index_ * parentNrow_]; }
+         inline reference operator*() { return start_[index_ * parentNrow_]; }
          
-         inline V* operator->() { return &(start_[index_ * parentNrow_]); }
+         inline pointer operator->() { return &(start_[index_ * parentNrow_]); }
       
-         inline V& operator[](int i) { return start_[(index_+i) * parentNrow_]; }
+         inline reference operator[](int i) { return start_[(index_+i) * parentNrow_]; }
          
       private:
-         V* start_;
-         std::size_t parentNrow_;
-         std::size_t index_;
+         pointer start_;
+         difference_type parentNrow_;
+         difference_type index_;
       };
       
       typedef row_iterator<T> iterator;

